### PR TITLE
⚠️ Use custom service accounts instead of default for controllers

### LINKED ISSUE #1206: Use custom service accounts instead of default for controllers
As an operator, I would like to use a dedicated service account for each CAPZ controller (instead of default)

/kind feature

/priority important-soon
/milestone v0.4.0
/help

see https://github.com/kubernetes-sigs/cluster-api/pull/4245

### PR DESCRIPTION
**Pull request by furkatgofurov7**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Uses a dedicated service account for each CAPZ controller (instead of default)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1206 

**Special notes for your reviewer**:

_Please have a look thoroughly, in case I missed some important points to update in the PR. Thanks!._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
⚠️ Use custom service accounts instead of default for controllers
```

### CODE REVIEW

**Comment by nprokopic** [comment]
Should the service account for aad-pod-identity be updated as well? https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/config/default/aad-pod-identity-deployment.yaml#L65-L67

**Comment by furkatgofurov7** [comment]
> Should the service account for aad-pod-identity be updated as well? https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/config/default/aad-pod-identity-deployment.yaml#L65-L67

@nprokopic hi! Thanks for the suggestion. I was mainly following [kubernetes-sigs/cluster-api/4245](https://github.com/kubernetes-sigs/cluster-api/pull/4245) and it changes only the SAs' in the `system` namespace, so I haven't touched it. But please do let me know if I am misunderstanding it.

**Comment by nprokopic** [comment]
> I was mainly following kubernetes-sigs/cluster-api/4245 and it changes only the SAs' in the system namespace

I believe that the changes are for the `default` service account in multiple namespaces with names ending with `-system` (there is some `kustomize` templating involved).

Here are the default service accounts from `*-system` namespaces before the above linked CAPI change:

```
$ kubectl get serviceaccount -A | grep default | grep system
capi-kubeadm-bootstrap-system       default                              1         28d
capi-kubeadm-control-plane-system   default                              1         28d
capi-system                         default                              1         28d
capi-webhook-system                 default                              1         28d
capz-system                         default                              1         28d
kube-system                         default                              1         28d
```

So I believe that in CAPZ we should also update the aad-pod-identity deployment which is currently using default service account for `capz-system` namespace, that is of course unless aad-pod-identity doesn't have some specific requirements and it needs the default service account.

**Comment by nader-ziada** [comment]
> So I believe that in CAPZ we should also update the aad-pod-identity deployment which is currently using default service account for `capz-system` namespace, that is of course unless aad-pod-identity doesn't have some specific requirements and it needs the default service account.

agree that aad-pod-identity service account should be updated as well

**Comment by furkatgofurov7** [comment]
@nprokopic thanks for spotting this up, fixed it.

**Comment by furkatgofurov7** [comment]
Seems a CI flakiness, re-triggering.

/retest

File: config/default/aad-pod-identity-deployment.yaml [hunk 1]
```diff
@@ -63,7 +63,7 @@ metadata:
     k8s-app: capz-aad-pod-id-nmi-binding
 subjects:
 - kind: ServiceAccount
-  name: default
+  name: manager
```
Comment (nprokopic) [inline config/default/aad-pod-identity-deployment.yaml:66]: The account that is set here I believe it should also be set in aad-pod-identity NMI DaemonSet `.Spec.Template.Spec.ServiceAccountName` field (in this same file), which is currently set to `default`.

Comment (furkatgofurov7) [reply config/default/aad-pod-identity-deployment.yaml:66]: Set it to `manager`. PTAL.

File: config/default/aad-pod-identity-deployment.yaml [hunk 2]
```diff
@@ -63,7 +63,7 @@ metadata:
...
   namespace: capz-system
```
Comment (nader-ziada) [inline config/default/aad-pod-identity-deployment.yaml:67]:
```suggestion
  namespace: system
```

the `manager` service account is created in `system` namespace

Comment (furkatgofurov7) [reply config/default/aad-pod-identity-deployment.yaml]: Fixed, thanks!

**Comment by nader-ziada** [comment]
/lgtm

@CecileRobertMichon PTAL

**Comment by CecileRobertMichon** [comment]
should this be marked as a breaking change? I see the corresponding CAPI commit is breaking

**Comment by furkatgofurov7** [comment]
@CecileRobertMichon done!

**Review by CecileRobertMichon — APPROVED** [review]
/approve

**Pull request merged**